### PR TITLE
Migrate to ZAF v2

### PR DESCRIPTION
**Pull request by fvzzy**
✌️ 

/cc @zendesk/vegemite 

### Description

- Migrates app over to ZAF v2
- Minor refactor to simplify testing
- Adds partial unit test coverage for app methods

7kzzl0enn6
wzxkstmj1w

### CODE REVIEW

**Comment by adammw** [comment]
probably don't need the banner or logo-promotion if it's just the default (especially since it's the old branding)

File: lib/templates/layout.hdbs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  <body>
+    <form class="search-wrapper">
+      <div class="input-append">
+        <input type="text" class="search-input" name="search-input">
+        <button class="btn search" type="button">
+          <i class="icon-search"></i>
+        </button>
+      </div>
+    </form>
+    <section data-main></section>
```
Comment (sandlerr) [inline lib/templates/layout.hdbs]: I'd have expected the form to go inside this element, or leave it out?

File: spec/helpers/zaf.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  return urlWithParams;
+}
+
+// Mock url params required by ZAF SDK
+if (history.pushState) {
+  let url = window.location.href;
+  let origin = encodeURIComponent(window.top.location.origin);
+  let app_guid = '36618c17-34c9-4ec3-8daa-24a9fa728461';
+  let newUrl = appendParams(url, `origin=${origin}&app_guid=${app_guid}`);
+  window.history.pushState({ path: newUrl }, '', newUrl);
```
Comment (adammw) [inline spec/helpers/zaf.js]: bit confused why we need this, was this just hard-coded for testing?

File: lib/templates/layout.hdbs (hunk 1 above)
Comment (sandlerr) [inline lib/templates/layout.hdbs]: never mind, the v1 app does the same thing

File: package.json [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  },
+  "license": "Apache-2.0",
+  "author": "Zendesk",
+  "keywords": [
+    "zendesk",
+    "app",
+    "framework"
+  ],
+  "bugs": {
+    "url": "https://github.com/zendesk/app_scaffold/issues"
```
Comment (sandlerr) [inline package.json]: fix these urls

File: src/javascripts/related_tickets_app.js [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+  },
+
+  onTicketSubjectChanged: _.debounce(function(ticketSubject) {
+    // don't search on empty subject lines
+    if (ticketSubject) {
+      const keywords = this.extractKeywords(ticketSubject);
+      const query = keywords.join(" ");
+
+      this.$('.search-input').val(query);
+      this.searchTickets(query);
```
Comment (adammw) [inline src/javascripts/related_tickets_app.js]: is it possible for a race condition to occur here, as in if a search for a previous query comes in after a subsequent search for a new query, the one that resolves last will clobber the correct result?

File: src/javascripts/related_tickets_app.js [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    // remove current ticket from results
+    if (currentTicketId) {
+      tickets = _.reject(tickets, function(ticket) {
+        return ticket.id === currentTicketId;
+      });
+    }
+
+    // trim the returned result string and append ellipses
+    _.each(tickets, function(ticket) {
+      ticket.description = ticket.description.substr(0,300).concat("...");
```
Comment (sandlerr) [inline src/javascripts/related_tickets_app.js]: nitpick: could replace the dots with this character: `…` or the html entity `&hellip;`

File: package.json [hunk 6]
```diff
@@ -0,0 +1,48 @@
+{
+  "name": "app_scaffold",
```
Comment (adammw) [inline package.json]: related_tickets_app ?

File: spec/related_tickets_app.spec.js [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+      searchInput.value = 'query string';
+      app.onSearchKeyPressed(e);
+      expect(app.searchTickets).toHaveBeenCalledWith('query string');
+    });
+
+    it('does not perform a search for an empty query string', () => {
+      const searchInput = document.createElement('input');
+      const e = {
+        target: searchInput,
+        which: 13,
```
Comment (sandlerr) [inline spec/related_tickets_app.spec.js]: maybe worth adding a testcase that it doesn't search if a different key was pressed?

**Comment by adammw** [comment]
For a PR like this, it's probably helpful to include a GIF showing the main functionality/changes in the description.

File: spec/related_tickets_app.spec.js [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+      };
+
+      searchInput.value = 'query string';
+      app.onSearchKeyPressed(e);
+      expect(app.searchTickets).toHaveBeenCalledWith('query string');
+    });
+
+    it('does not perform a search for an empty query string', () => {
+      const searchInput = document.createElement('input');
+      const e = {
```
Comment (adammw) [inline spec/related_tickets_app.spec.js]: Could potentially DRY up this event and move it up a scope / Object.freeze it

**Comment by sandlerr** [comment]
Looking great! Updated screenshots in the repo and a gif in the pr would be awesome.

File: spec/helpers/zaf.js (hunk 2 above)
Comment (danielbreves) [inline spec/helpers/zaf.js]: Yes

File: src/javascripts/related_tickets_app.js [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+    'search.fail'             : 'onSearchFailed'
+  },
+
+  init: function() {
+    const client = this.zafClient;
+
+    client.get('ticket').then((ticket) => {
+      this.ticket = {};
+      this.ticket.id = ticket.ticket.id;
+      this.ticket.subject = ticket.ticket.subject;
```
Comment (danielbreves) [inline src/javascripts/related_tickets_app.js]: This may cause a race condition if a function tries to read `this.ticket` before the promise here resolves.

File: templates/layout.hdbs [hunk 10]
```diff
@@ -1,15 +0,0 @@
-<header>
-  <span class="logo"/>
-  <h3>{{setting "title"}}</h3>
-</header>
-<form class="search-wrapper">
-  <div class="row-fluid">
-    <div class="span11">
-      <input type="text" placeholder="{{t "search.keywords"}}" class="search-input" name="search-input">
```
Comment (danielbreves) [inline templates/layout.hdbs]: Could we set `placeholder` dynamically once the app loads?

**Comment by danielbreves** [comment]
Nice work!

File: package.json [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+  "description": "A Zendesk app that shows a list of tickets related to the current ticket being viewed.",
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/zendesk/related_tickets_app.git"
+  },
+  "scripts": {
+    "build": "webpack -p",
+    "watch": "webpack --watch"
+  },
+  "license": "MIT",
```
Comment (danielbreves) [inline package.json]: License should actually be Apache 2.0. See https://zendesk.atlassian.net/wiki/display/ENG/Open+Source.

Comment (sandlerr) [inline package.json]: doesn't this conflict with the LICENSE file?

File: dist/manifest.json [hunk 12]
```diff
@@ -3,12 +3,16 @@
...
+      "ticket_sidebar": "assets/index.html",
+      "new_ticket_sidebar": "assets/index.html"
+    }
+  },
   "defaultLocale": "en",
   "private": false,
   "singleInstall": true,
-  "frameworkVersion": "1.0",
+  "frameworkVersion": "2.0",
   "version": "1.0.1",
```
Comment (sandlerr) [inline dist/manifest.json]: let's make sure we bump this and tag the repo before shipping (probably after merging though)

File: src/javascripts/related_tickets_app.js [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+    'keydown .search-input'   : 'onSearchKeyPressed',
+    'click .search'           : 'onSearchClicked',
+    'search.done'             : 'onSearchDone',
+    'search.fail'             : 'onSearchFailed'
+  },
+
+  init: function() {
+    const client = this.zafClient;
+
+    client.get('ticket').then((ticket) => {
```
Comment (sandlerr) [inline src/javascripts/related_tickets_app.js]: redundant brackets :)

File: src/javascripts/related_tickets_app.js [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+    if (!_.isEmpty(query)) {
+      this.searchTickets(query);
+    }
+  },
+
+  onSearchDone: function(resultsObj) {
+    const client = this.zafClient;
+    let currentTicketId;
+
+    client.get('ticket').then((ticket) => {
```
Comment (sandlerr) [inline src/javascripts/related_tickets_app.js]: let's not re-fetch the ticket, you did it already in `init` and the ID won't change.

**Comment by sandlerr** [comment]
We're going to need to change the deploy script as well, let's talk about that and get it working ideally before merging

File: src/javascripts/related_tickets_app.js (hunk 14 above)
Comment (danielbreves) [inline src/javascripts/related_tickets_app.js]: @sandlerr there's a small chance this event could fire before the initial fetch in `init` so the best we could do would be to write a function that caches the `ticket.id` after the first fetch and returns a resolved promise, but not sure it's worth the trouble here. We should probably change this to get the `ticket.id` instead of the `ticket` though.

File: karma.conf.js [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+
+    customLaunchers: {
+      // CRUFT: needed to load zaf_sdk with crossorigin=anonymous
+      Chrome_no_security: {
+        base: 'Chrome',
+        flags: ['--disable-web-security']
+      },
+      Chrome_travis_ci: {
+        base: 'Chrome_no_security',
+        flags: ['--no-sandbox']
```
Comment (sandlerr) [inline karma.conf.js]: odd, why did you need this?

Comment (fvzzy) [inline karma.conf.js]: I actually didn't test it without this flag but assumed it was required based on the end of this thread: https://github.com/karma-runner/karma/issues/1144. I'll try without and see what happens.
